app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt

```diff
@@ -659,18 +648,61 @@ class ParkingViewModel(
 
   // ================== Images ==================
 
-  companion object {
-    val Factory: ViewModelProvider.Factory =
-        object : ViewModelProvider.Factory {
-          @Suppress("UNCHECKED_CAST")
-          override fun <T : ViewModel> create(modelClass: Class<T>): T {
-            return ParkingViewModel(
-                ImageRepositoryCloudStorage(
-                    FirebaseAuth.getInstance(), FirebaseStorage.getInstance()),
-                ParkingRepositoryFirestore(FirebaseFirestore.getInstance()),
-                ReportedObjectRepositoryFirestore(FirebaseFirestore.getInstance()))
-                as T
-          }
-        }
+  // ================== Offline ==================
+
+  /**
+   * Downloads the parkings in the zone to download from the local storage.
+   *
+   * @param zone the zone to download
+   * @param onSuccess the callback function to execute when the download is successful
+   * @param onFailure the callback function to execute when the download fails
+   */
+  fun downloadZone(zone: Zone, onSuccess: () -> Unit, onFailure: (Exception) -> Unit) {
+    val tilesToDownload =
+        TileUtils.getAllTilesInRectangle(zone.boundingBox.southwest(), zone.boundingBox.northeast())
+
+    tilesToDownload.forEach { tile ->
+      if (tilesToParking.containsKey(tile)) {
+        offlineParkingRepository.downloadParkings(tilesToParking[tile]!!) {}
+      } else {
+        parkingRepository.getParkingsForTile(
+            tile,
+            {
+              offlineParkingRepository.downloadParkings(it) {
+                Log.d("ParkingViewModel", "Tile downloaded successfully")
+              }
+            },
+            {
+              Log.e("ParkingViewModel", "Error getting parkings for tile: $it")
+              onFailure(it)
+            })
+      }
+    }
+    onSuccess()
+  }
+
+  /**
+   * Deletes the parkings in the zone to delete from the local storage.
+   *
+   * @param zoneToDelete the zone to delete
+   * @param allZones the list of all zones
+   */
+  fun deleteZone(zoneToDelete: Zone, allZones: List<Zone>) {
+    val tilesToKeep =
+        allZones
+            .flatMap { zone ->
+              if (zone != zoneToDelete)
+                  TileUtils.getAllTilesInRectangle(
+                      zone.boundingBox.southwest(), zone.boundingBox.northeast())
+              else emptySet()
+            }
+            .toSet()
+
+    val tilesToDelete =
+        TileUtils.getAllTilesInRectangle(
+            zoneToDelete.boundingBox.southwest(), zoneToDelete.boundingBox.northeast())
+    offlineParkingRepository.deleteTiles(tilesToDelete - tilesToKeep) {
```

Comment:
Again, a quick error handle to check that tilesToDelete > tilesToKeep would be nice

Reply:
Since those are sets, the difference can only be the empty set in the worst case :)

Reply:
Good point!